### PR TITLE
[ENH] Increase AptaTrans test coverage

### LINKED ISSUE #180: AptaTrans test coverage
Following #177, we should review test coverage of AptaTrans to make sure everything works correctly. 

Below, the test coverage for AptaTrans related modules.

```
=============================== tests coverage ================================
_______________ coverage: platform win32, python 3.12.9-final-0 _______________

Name                                                            Stmts   Miss Branch BrPart  Cover
-------------------------------------------------------------------------------------------------
pyaptamer\aptatrans\__init__.py                                     6      0      0      0   100%
pyaptamer\aptatrans\_model.py                                      80     10     14      2    85%
pyaptamer\aptatrans\_model_lightning.py                            34      0      0      0   100%
pyaptamer\aptatrans\_pipeline.py                                   42      0      6      1    98%
pyaptamer\aptatrans\layers\__init__.py                              3      0      0      0   100%
pyaptamer\aptatrans\layers\_convolutional.py                       24      0      4      0   100%
pyaptamer\aptatrans\layers\_encoder.py                             38      0      2      1    98%
pyaptamer\aptatrans\layers\_interaction_map.py                     14      0      0      0   100%
pyaptamer\aptatrans\layers\tests\__init__.py                        1      0      0      0   100%
pyaptamer\aptatrans\layers\tests\test_all_aptatrans_layers.py      78      0      2      0   100%
pyaptamer\aptatrans\tests\__init__.py                               1      0      0      0   100%
pyaptamer\aptatrans\tests\test_aptatrans.py                       140      4      0      0    97%
pyaptamer\aptatrans\tests\test_aptatrans_lightning.py              66      0      0      0   100%
```

Coverage looks good overall. Only `_model.py` is relatively low, however the branch not being tested is the one related to loading pretraining weights, not crucial.

As I mentioned in #178, the bug on shape mismatch was caused by introducing the attention mask in the encoders. This wasn't something we could have catched via test as it's, rather than a bug, was an non-conventional way of doing things (i.e., not using an attention mask).

FYI @fkiraly

### PR DESCRIPTION
**Pull request by goyaladitya05**
#### Reference Issues/PRs
Closes: #180 

#### What does this implement/fix? Explain your changes.
Added tests to improve coverage of AptaTrans-related modules:

- `test_positional_encoding_exceeds_max_len` - covers the AssertionError branch in PositionalEncoding.forward() when input exceeds max_len
- `test_forward_encoder_invalid_type` - covers the ValueError branch in forward_encoder() for unknown encoder_type
- `test_forward_encoder_with_padding` - covers the padding-mask branch in _make_encoder(), parametrized over inputs with and without zero-padding
- `test_load_pretrained_weights_local` / `test_load_pretrained_weights_remote` - covers both branches of load_pretrained_weights() (local file and hf download fallback)
- `test_recommend_verbose` - covers the verbose print path in AptaTransPipeline.recommend()

#### Did you add any tests for the change?

Yes, all changes are new tests.

#### PR checklist

- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.
- [x] Added/modified tests
- [x] Used pre-commit hooks when committing to ensure that code is compliant with hooks. Install hooks with `pre-commit install`.
  To run hooks independent of commit, execute `pre-commit run --all-files`

## Before vs After test results

| Module | Stmts | Miss (before) | Cover (before) | Miss (after) | Cover (after) |
|---|---|---|---|---|---|
| `aptatrans/__init__.py` | 6 | 0 | 100% | 0 | 100% |
| `aptatrans/_model.py` | 82 | 10 | 88% | 1 | **99%** |
| `aptatrans/_model_lightning.py` | 54 | 6 | 89% | 6 | 89% |
| `aptatrans/_pipeline.py` | 43 | 0 | 100% | 0 | 100% |
| `layers/_convolutional.py` | 24 | 0 | 100% | 0 | 100% |
| `layers/_encoder.py` | 38 | 0 | 100% | 0 | 100% |
| `layers/_interaction_map.py` | 14 | 0 | 100% | 0 | 100% |
| `layers/tests/test_all_aptatrans_layers.py` | 78->85 | 0 | 100% | 0 | 100% |
| `tests/test_aptatrans.py` | 146->200 | 4 | 97% | 4 | 98% |
| `tests/test_aptatrans_lightning.py` | 70 | 0 | 100% | 0 | 100% |
| **TOTAL** | **560->621** | **20** | **96%** | **11** | **98%** |

**57 passed, 1 skipped** (before) -> **65 passed, 1 skipped** (after)

### CODE REVIEW

**Comment by siddharth7113** [comment]
Hi,

Please add before and after test coverage results here, since it would be easy to understand the effect of the PR.

**Comment by goyaladitya05** [comment]
> Please add before and after test coverage results here, since it would be easy to understand the effect of the PR.

@siddharth7113 updated.

**Comment by goyaladitya05** [comment]
Hi @NennoMP @siddharth7113, could you please review? thanks.

File: pyaptamer/aptatrans/tests/test_aptatrans.py [hunk 1]
```diff
@@ -2,6 +2,9 @@
 
 __author__ = ["nennomp"]
 
+from pathlib import Path
+from unittest.mock import patch
+
```
Comment (NennoMP) [inline pyaptamer/aptatrans/tests/test_aptatrans.py]: The codebase uses `pytest`. Avoid introducing an additional dependency such as `unittest`.

File: pyaptamer/aptatrans/tests/test_aptatrans.py [hunk 2]
```diff
@@ -470,3 +596,26 @@ def test_get_interaction_map(self, device, candidate, target):
             model.apta_embedding.max_len,
             model.prot_embedding.max_len,
         )
+
+    def test_recommend_verbose(self, monkeypatch, capsys):
```
Comment (NennoMP) [inline pyaptamer/aptatrans/tests/test_aptatrans.py]: I don't think we need an entire new test just to cover the case of `verbose=True`. I would remove this or, at least, find a way to include it in `test_recommend()` via `pytest.mark.parametrize` or something similar.

**Review by NennoMP — CHANGES_REQUESTED** [review]

File: pyaptamer/aptatrans/tests/test_aptatrans.py (hunk 2 above)
Comment (goyaladitya05) [reply pyaptamer/aptatrans/tests/test_aptatrans.py]: removed that and added verbose=True as a third parametrize case in test_recommend instead

File: pyaptamer/aptatrans/tests/test_aptatrans.py [hunk 3]
```diff
@@ -202,6 +202,130 @@ def test_forward_output_shape_asymmetric(
...
+        self.counter = 0
+
+    def run(self, verbose: bool = False):
+        result = {
+            "candidate": f"APTAMER_{self.counter:03d}",
+            "sequence": f"seq_{self.counter}",
+            "score": torch.tensor(0.75),
+        }
+        self.counter += 1
+        return result
```
Comment (NennoMP) [inline pyaptamer/aptatrans/tests/test_aptatrans.py]: Any reason this is being added? We already have a MockMCTS inside `test_recommend()` which is the only method actually using it. In any case, currently this is essentially a duplicate mock,

**Review by NennoMP — CHANGES_REQUESTED** [review]

File: pyaptamer/aptatrans/tests/test_aptatrans.py (hunk 3 above)
Comment (goyaladitya05) [reply pyaptamer/aptatrans/tests/test_aptatrans.py]: you're right. It had relevence with the changes I had previously, but as of now its a duplicate. removed.

**Comment by NennoMP** [comment]
Tests are failing, please ensure that all checks are successful.

**Comment by goyaladitya05** [comment]
Closing, as I'm not able to work on this for now.
